### PR TITLE
WIP Add upload to azure functionality

### PR DESCRIPTION
**Pull request by Qi77Qi**
TODO:
Figure out how to use managed identity instead of access key for auth.

Tested on a terra VM

```
>>> drs.copy("drs://jade.datarepo-dev.broadinstitute.org/v1_0c86170e-312d-4b39-a0a4-2a2bfaa24c7a_c0e40912-8b14-43f6-9a2f-b278144d0060", "https://qijlbdgpc4zqdee.blob.core.windows.net/qi-test-container/subdir/jade1")
2021-09-22 05:50:19::INFO  Enabling requester pays for your workspace. This will only take a few seconds...
2021-09-22 05:50:19::WARNING  Failed to init requester pays for workspace qi-monitoring-1218/qi-ws-1: Expected '204', got '401' for 'https://rawls.dsde-prod.broadinstitute.org/api/workspaces/qi-monitoring-1218/qi-ws-1/enableRequesterPaysForLinkedServiceAccounts'. You will not be able to access DRS URIs that interact with requester pays buckets.
2021-09-22 05:50:21::INFO  Request URL: 'https://qijlbdgpc4zqdee.blob.core.windows.net/qi-test-container/subdir/jade1'
2021-09-22 05:50:21::INFO  Request method: 'PUT'
2021-09-22 05:50:21::INFO  Request headers:
2021-09-22 05:50:21::INFO      'x-ms-blob-type': 'REDACTED'
2021-09-22 05:50:21::INFO      'Content-Length': '62043448'
2021-09-22 05:50:21::INFO      'If-None-Match': '*'
2021-09-22 05:50:21::INFO      'x-ms-version': 'REDACTED'
2021-09-22 05:50:21::INFO      'Content-Type': 'application/octet-stream'
2021-09-22 05:50:21::INFO      'Accept': 'application/xml'
2021-09-22 05:50:21::INFO      'User-Agent': 'azsdk-python-storage-blob/12.9.0 Python/3.7.10 (Linux-5.4.104+-x86_64-with-debian-buster-sid)'
2021-09-22 05:50:21::INFO      'x-ms-date': 'REDACTED'
2021-09-22 05:50:21::INFO      'x-ms-client-request-id': '8e96e9a6-1bcd-11ec-b6fa-0242ac120005'
2021-09-22 05:50:21::INFO      'Authorization': 'REDACTED'
2021-09-22 05:50:21::INFO  A body is sent with the request
2021-09-22 05:50:23::INFO  Response status: 201
2021-09-22 05:50:23::INFO  Response headers:
2021-09-22 05:50:23::INFO      'Content-Length': '0'
2021-09-22 05:50:23::INFO      'Content-MD5': 'REDACTED'
2021-09-22 05:50:23::INFO      'Last-Modified': 'Wed, 22 Sep 2021 17:50:23 GMT'
2021-09-22 05:50:23::INFO      'ETag': '"0x8D97DF173D92DB6"'
2021-09-22 05:50:23::INFO      'Server': 'Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0'
2021-09-22 05:50:23::INFO      'x-ms-request-id': '565e8615-201e-00d6-41da-af7108000000'
2021-09-22 05:50:23::INFO      'x-ms-client-request-id': '8e96e9a6-1bcd-11ec-b6fa-0242ac120005'
2021-09-22 05:50:23::INFO      'x-ms-version': 'REDACTED'
2021-09-22 05:50:23::INFO      'x-ms-content-crc64': 'REDACTED'
2021-09-22 05:50:23::INFO      'x-ms-request-server-encrypted': 'REDACTED'
2021-09-22 05:50:23::INFO      'Date': 'Wed, 22 Sep 2021 17:50:23 GMT'
 DefaultEndpointsProtocol=https;AccountNa   100%   [========================================]   59.2MiB   17.9MiB/s   3.31s
```

### CODE REVIEW

File: tests/test_http.py [hunk 1]
```diff
@@ -10,7 +10,7 @@
 
 from tests import config
 from tests.infra.server import ThreadedLocalServer, BaseHTTPRequestHandler
-from terra_notebook_utils.http import HTTPAdapter, Retry, http_session
+from terra_notebook_utils.http_session import HTTPAdapter, Retry, http_session
```
Comment (Qi77Qi) [inline tests/test_http.py:13]: I had to rename `http` file because of some naming conflict error

File: README.md [hunk 2]
```diff
@@ -169,18 +169,21 @@ For local development:
...
-2. run `docker pull us.gcr.io/broad-dsp-gcr-public/terra-jupyter-python:0.0.12`
-3. run the image from *one directory above* the root directory of this repo via `docker run -itd --entrypoint='/bin/bash' -v $PWD/terra-notebook-utils:/work -u root -e PIP_USER=false --name test-image us.gcr.io/broad-dsp-gcr-public/terra-jupyter-python:0.0.12`
+2. run `docker pull us.gcr.io/broad-dsp-gcr-public/terra-jupyter-python:1.0.1`
+3. run the image from *one directory above* the root directory of this repo via `docker run -itd --entrypoint='/bin/bash' -v $PWD/terra-notebook-utils:/work -u root -e PIP_USER=false --name test-image us.gcr.io/broad-dsp-gcr-public/terra-jupyter-python:1.0.1`
 4. Attach your terminal to the image via `docker exec -it test-image bash`, then navigate to the directory the code is mounted to via `cd /work`. Note that the above command ensures any changes you make to files in the repo will be updated in the image as well.
 5. log in with your Google credentials using `gcloud auth application-default login`,
-6. install requirements with `pip install -r requirements.txt`
+6. install requirements with `pip3 install -r requirements.txt`
```
Comment (xbrianh) [inline README.md:176]: Python developers typically work in Python [virtual environments](https://docs.python.org/3/tutorial/venv.html). Inside a Python 3 virtual environment you use `pip` and `python`, not `pip3` and `python3`.

This line can be reverted, and we can rely on developers to understand which `pip` to use.

File: terra_notebook_utils/blobstore/azure_blob_store.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+from terra_notebook_utils import blobstore
+from azure.storage.blob import BlobServiceClient
+import os
+from azure.identity import DefaultAzureCredential
+
+class AzureBlobStore(blobstore.BlobStore):
+    schema = "https://"
```
Comment (xbrianh) [inline terra_notebook_utils/blobstore/azure_blob_store.py:7]: This schema is unfortunately awkward for TNU. Currently, the CLI command to copy a drs file into a Google bucket is
```
tnu drs copy drs://foo gs://my-bucket/my-key
```

Note the `gs://` schema. The analogous command for an azure container looks awkward due to the `https://` schema
```
tnu drs copy drs://foo https://some-weird-azure-url
```

A further complication is URL detection. TNU uses the schema to understand the storage provider for destination URLs. Logic for detecting Azure destinations will need to be added [here](https://github.com/DataBiosphere/terra-notebook-utils/blob/f326afb538da730f0f54ad015c125323a64065f1/terra_notebook_utils/blobstore/copy_client.py#L91-L101).

File: terra_notebook_utils/blobstore/azure_blob_store.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+    def size(self) -> int:
+        blob = self._azure_blob_client.get_blob_properties()
+        return blob.size
+
+    def delete(self):
+        self._azure_blob_client.delete_blob("include")
+
+    def exists(self):
+        return self._azure_blob_client.exists()
```
Comment (xbrianh) [inline terra_notebook_utils/blobstore/azure_blob_store.py:60]: Multipart uploads are supported for both [s3](https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html) and [gs](https://cloud.google.com/storage/docs/composite-objects). It is typically more performant to upload large objects as parts, sometimes concurrently. Also, it my not be possible to upload a large object as a single part. For instance in S3 you cannot upload an object larger than 5GB with a single PUT.

How are large object uploads handled in Azure?

File: tests/test_http.py (hunk 1 above)
Comment (xbrianh) [reply tests/test_http.py:13]: What was the conflict?

**Review by xbrianh — COMMENTED** [review]
This looks like good progress, and it's exciting to see TNU support Azure storage :)

I have left a few comments and questions. Also, the test suites for [blobstore](https://github.com/DataBiosphere/terra-notebook-utils/blob/master/tests/test_blobstore.py) and [copy_client](https://github.com/DataBiosphere/terra-notebook-utils/blob/master/tests/test_copy_client.py) will need to be extended to cover Azure operations.

File: terra_notebook_utils/blobstore/azure_blob_store.py (hunk 4 above)
Comment (Qi77Qi) [reply terra_notebook_utils/blobstore/azure_blob_store.py:60]: it seems like multi part is supported under the hood by the azure sdk...see [this](https://github.com/Azure/azure-sdk-for-python/issues/16190)

**Comment by Qi77Qi** [comment]
@xbrianh the error was something like `no module found: http`..and I googled a bit for the error, and someone suggested it could be naming conflict, so I changed the name of that file

**Comment by Qi77Qi** [comment]
closing this PR in favor of https://github.com/DataBiosphere/terra-notebook-utils/pull/362 since this is from a fork

**Pull request closed without merging**